## Learning Pandas & apply PCA technique for 2D visualization

In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 5GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/pokemon/Pokemon.csv


### Load data file into pandas dataframe

In [2]:
df = pd.read_csv('../input/pokemon/Pokemon.csv')
df.head(3)

#       Name Type 1  Type 2  Total  HP  Attack  Defense  Sp. Atk  Sp. Def  \
0  1  Bulbasaur  Grass  Poison    318  45      49       49       65       65   
1  2    Ivysaur  Grass  Poison    405  60      62       63       80       80   
2  3   Venusaur  Grass  Poison    525  80      82       83      100      100   

   Speed  Generation  Legendary  
0     45           1      False  
1     60           1      False  
2     80           1      False

#### Read summary of a dataframe including dtype and non-null value count.

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800 entries, 0 to 799
Data columns (total 13 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   #           800 non-null    int64 
 1   Name        800 non-null    object
 2   Type 1      800 non-null    object
 3   Type 2      414 non-null    object
 4   Total       800 non-null    int64 
 5   HP          800 non-null    int64 
 6   Attack      800 non-null    int64 
 7   Defense     800 non-null    int64 
 8   Sp. Atk     800 non-null    int64 
 9   Sp. Def     800 non-null    int64 
 10  Speed       800 non-null    int64 
 11  Generation  800 non-null    int64 
 12  Legendary   800 non-null    bool  
dtypes: bool(1), int64(9), object(3)
memory usage: 75.9+ KB


#### Shape function tells shape or dimension of dataframe i.e no. of rows & columns.


In [4]:
df.shape

(800, 13)

#### Read headers i.e column name


In [5]:
df.columns

Index(['#', 'Name', 'Type 1', 'Type 2', 'Total', 'HP', 'Attack', 'Defense',
       'Sp. Atk', 'Sp. Def', 'Speed', 'Generation', 'Legendary'],
      dtype='object')

#### Change column name '#' to 'id'


In [6]:
columns = df.columns.tolist()
columns[0] = 'id'
df.columns = columns
df.columns

Index(['id', 'Name', 'Type 1', 'Type 2', 'Total', 'HP', 'Attack', 'Defense',
       'Sp. Atk', 'Sp. Def', 'Speed', 'Generation', 'Legendary'],
      dtype='object')

#### Read any particular column.

In [7]:
df['Name'][0:5]

##Read more than one columns
df[['Name', 'Attack','Type 2']][0:4]

Name  Attack  Type 2
0              Bulbasaur      49  Poison
1                Ivysaur      62  Poison
2               Venusaur      82  Poison
3  VenusaurMega Venusaur     100  Poison

#### Read more than 1 column at a time and store in new dataframe.

In [8]:
df_tmp = df[['Name', 'Type 1', 'Attack']]
df_tmp.head(2)

Name Type 1  Attack
0  Bulbasaur  Grass      49
1    Ivysaur  Grass      62

#### Read any row using iloc function using its index

In [9]:
df.iloc[1]

id                  2
Name          Ivysaur
Type 1          Grass
Type 2         Poison
Total             405
HP                 60
Attack             62
Defense            63
Sp. Atk            80
Sp. Def            80
Speed              60
Generation          1
Legendary       False
Name: 1, dtype: object

#### Read a value from specific row & column

In [10]:
df.iloc[3,2]

'Grass'

#### Iterate through each row

In [11]:
# for index, row in df.iterrows():
#     ##print(index, row)
#     print(index, row['Name'])

#### Apply conditional statements and read only those who satisfies

In [12]:
df.loc[df['Type 1'] == 'Fire'][0:3]

id        Name Type 1  Type 2  Total  HP  Attack  Defense  Sp. Atk  \
4   4  Charmander   Fire     NaN    309  39      52       43       60   
5   5  Charmeleon   Fire     NaN    405  58      64       58       80   
6   6   Charizard   Fire  Flying    534  78      84       78      109   

   Sp. Def  Speed  Generation  Legendary  
4       50     65           1      False  
5       65     80           1      False  
6       85    100           1      False

### Sorting data with respect to a particular/specify column 

In [13]:
##Sort dataframe in ascending order(default) through a particular column.
df.sort_values('Attack')

##Sort dataframe in descending order.
df.sort_values('Attack', ascending =False)

##Sort dataframe using multiple columns with different order
df.sort_values(['Attack', 'Defense'], ascending =(0,1))[:5]

id                     Name   Type 1    Type 2  Total   HP  Attack  \
163  150      MewtwoMega Mewtwo X  Psychic  Fighting    780  106     190   
232  214  HeracrossMega Heracross      Bug  Fighting    600   80     185   
429  386       DeoxysAttack Forme  Psychic       NaN    600   50     180   
426  384    RayquazaMega Rayquaza   Dragon    Flying    780  105     180   
424  383    GroudonPrimal Groudon   Ground      Fire    770  100     180   

     Defense  Sp. Atk  Sp. Def  Speed  Generation  Legendary  
163      100      154      100    130           1       True  
232      115       40      105     75           2      False  
429       20      180       20    150           3       True  
426      100      180      100    115           3       True  
424      160      150       90     90           3       True

### Data Manipulation :- Adding new column in dataframe 


In [14]:
df['Total'] = df['HP'] + df['Attack'] + df['Defense'] + df['Sp. Atk'] +df['Sp. Def'] +df['Speed']
df.head(3)

##Add new column using iloc[row, column].sum(axis = 1 for horizontal sum & 0 for vertical sum) 
df['Total_iloc_use'] = df.iloc[:, 4:10].sum(axis = 1)
df.head(3)

id       Name Type 1  Type 2  Total  HP  Attack  Defense  Sp. Atk  Sp. Def  \
0   1  Bulbasaur  Grass  Poison    318  45      49       49       65       65   
1   2    Ivysaur  Grass  Poison    405  60      62       63       80       80   
2   3   Venusaur  Grass  Poison    525  80      82       83      100      100   

   Speed  Generation  Legendary  Total_iloc_use  
0     45           1      False             591  
1     60           1      False             750  
2     80           1      False             970

#### Drop columns in a dataframe

In [15]:
df = df.drop(columns =['Total_iloc_use'])
df[:3]

id       Name Type 1  Type 2  Total  HP  Attack  Defense  Sp. Atk  Sp. Def  \
0   1  Bulbasaur  Grass  Poison    318  45      49       49       65       65   
1   2    Ivysaur  Grass  Poison    405  60      62       63       80       80   
2   3   Venusaur  Grass  Poison    525  80      82       83      100      100   

   Speed  Generation  Legendary  
0     45           1      False  
1     60           1      False  
2     80           1      False

### Save modified data in new data file(.csv, .xlsx)

In [16]:
##Save your datsframe as new csv file without index
df.to_csv('modified_new.csv', index=False)

##Save in excel format i.e xlsx
df.to_excel('modified_new_excel.xlsx', index=False)

### Filtering Dataframe

In [17]:
df.loc[(df['Type 1'] =='Grass') & (df['Type 2']=='Poison') & (df['Attack'] >=69)][:4]

##Save this filtered data in new dataframe
new_df = df.loc[(df['Type 1'] =='Grass') & (df['Type 2']=='Poison') & (df['Attack'] >=69)][:4]

##new_df.head(3)

##Reset index of new dataframe
new_df.reset_index(drop =True, inplace=True)
new_df.head(3)

id                   Name Type 1  Type 2  Total  HP  Attack  Defense  \
0   3               Venusaur  Grass  Poison    525  80      82       83   
1   3  VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
2  45              Vileplume  Grass  Poison    490  75      80       85   

   Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0      100      100     80           1      False  
1      122      120     80           1      False  
2      110       90     50           1      False

In [18]:
##Find value in 'Name' column which contains mega word

df.loc[df['Name'].str.contains('Mega')][:4]

##Remove those Name from data which contains mega word  

df.loc[~df['Name'].str.contains('Mega')]

id                 Name   Type 1  Type 2  Total   HP  Attack  Defense  \
0      1            Bulbasaur    Grass  Poison    318   45      49       49   
1      2              Ivysaur    Grass  Poison    405   60      62       63   
2      3             Venusaur    Grass  Poison    525   80      82       83   
4      4           Charmander     Fire     NaN    309   39      52       43   
5      5           Charmeleon     Fire     NaN    405   58      64       58   
..   ...                  ...      ...     ...    ...  ...     ...      ...   
794  718     Zygarde50% Forme   Dragon  Ground    600  108     100      121   
795  719              Diancie     Rock   Fairy    600   50     100      150   
797  720  HoopaHoopa Confined  Psychic   Ghost    600   80     110       60   
798  720   HoopaHoopa Unbound  Psychic    Dark    680   80     160       60   
799  721            Volcanion     Fire   Water    600   80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  
0         65       65     45           1      False  
1         80       80     60           1      False  
2        100      100     80           1      False  
4         60       50     65           1      False  
5         80       65     80           1      False  
..       ...      ...    ...         ...        ...  
794       81       95     95           6       True  
795      100      150     50           6       True  
797      150      130     70           6       True  
798      170      130     80           6       True  
799      130       90     70           6       True  

[751 rows x 13 columns]

### Regex library for further filtering operations

In [19]:
import re

##Find 'Name' which starts with 'pi' in Name column

df.loc[df['Name'].str.contains('^pi[a-z]*', flags =re.I, regex= True)]

id                 Name    Type 1    Type 2  Total   HP  Attack  \
20    16               Pidgey    Normal    Flying    251   40      45   
21    17            Pidgeotto    Normal    Flying    349   63      60   
22    18              Pidgeot    Normal    Flying    479   83      80   
23    18  PidgeotMega Pidgeot    Normal    Flying    579   83      80   
30    25              Pikachu  Electric       NaN    320   35      55   
136  127               Pinsir       Bug       NaN    500   65     125   
137  127    PinsirMega Pinsir       Bug    Flying    600   65     155   
186  172                Pichu  Electric       NaN    205   20      40   
219  204               Pineco       Bug       NaN    290   50      65   
239  221            Piloswine       Ice    Ground    450  100     100   
438  393               Piplup     Water       NaN    314   53      51   
558  499              Pignite      Fire  Fighting    418   90      93   
578  519               Pidove    Normal    Flying    264   50      55   

     Defense  Sp. Atk  Sp. Def  Speed  Generation  Legendary  
20        40       35       35     56           1      False  
21        55       50       50     71           1      False  
22        75       70       70    101           1      False  
23        80      135       80    121           1      False  
30        40       50       50     90           1      False  
136      100       55       70     85           1      False  
137      120       65       90    105           1      False  
186       15       35       35     60           2      False  
219       90       35       35     15           2      False  
239       80       60       60     50           2      False  
438       53       61       56     40           4      False  
558       55       70       55     55           5      False  
578       50       36       30     43           5      False

### Conditional changes in dataframe

In [20]:
##Changing the value 'Fire' to 'Flamer' in the 'Name' column 

df.loc[df['Type 1'] =='Fire', 'Type 1'] ='Flamer'
df

##Make all fire type pokemon legendery
df.loc[df['Type 1'] =='Flamer', 'Legendary'] =True
df

##Add new column and assign true value whose attack value is greater than 69
df.loc[df['Attack'] > 69, 'Beast'] = True
df.loc[df['Attack'] < 69, 'Beast'] = False
df

id                   Name   Type 1  Type 2  Total  HP  Attack  Defense  \
0      1              Bulbasaur    Grass  Poison    318  45      49       49   
1      2                Ivysaur    Grass  Poison    405  60      62       63   
2      3               Venusaur    Grass  Poison    525  80      82       83   
3      3  VenusaurMega Venusaur    Grass  Poison    625  80     100      123   
4      4             Charmander   Flamer     NaN    309  39      52       43   
..   ...                    ...      ...     ...    ...  ..     ...      ...   
795  719                Diancie     Rock   Fairy    600  50     100      150   
796  719    DiancieMega Diancie     Rock   Fairy    700  50     160      110   
797  720    HoopaHoopa Confined  Psychic   Ghost    600  80     110       60   
798  720     HoopaHoopa Unbound  Psychic    Dark    680  80     160       60   
799  721              Volcanion   Flamer   Water    600  80     110      120   

     Sp. Atk  Sp. Def  Speed  Generation  Legendary  Beast  
0         65       65     45           1      False  False  
1         80       80     60           1      False  False  
2        100      100     80           1      False   True  
3        122      120     80           1      False   True  
4         60       50     65           1       True  False  
..       ...      ...    ...         ...        ...    ...  
795      100      150     50           6       True   True  
796      160      110    110           6       True   True  
797      150      130     70           6       True   True  
798      170      130     80           6       True   True  
799      130       90     70           6       True   True  

[800 rows x 14 columns]

#### Read modified data

In [21]:
df_mod = pd.read_csv('modified_new.csv')
df_mod.head(3)

id       Name Type 1  Type 2  Total  HP  Attack  Defense  Sp. Atk  Sp. Def  \
0   1  Bulbasaur  Grass  Poison    318  45      49       49       65       65   
1   2    Ivysaur  Grass  Poison    405  60      62       63       80       80   
2   3   Venusaur  Grass  Poison    525  80      82       83      100      100   

   Speed  Generation  Legendary  
0     45           1      False  
1     60           1      False  
2     80           1      False

### Aggregate functions i.e groupby(), sum(), count() and mean()

In [22]:
##Aggregate Statistics will be done with groupby(), sum(), count() and mean() function.

##Use groupby function to find mean of all Type 1 pokemon and sort in ascending order w.r.t Attack.

df_mod.groupby(['Type 1']).mean().sort_values('Attack', ascending=False)

##Sum up all Type 1 pokemon features
df_mod.groupby(['Type 1']).sum()

##Count number of Type 1 pokemon in dataframe
df_mod.groupby(['Type 1']).count()

##Make another column count to count Type 1 pokemon more efficiently &  it is helpful in big data 
df_mod['count'] =1

df_mod.groupby(['Type 1']).count()['count']

##Apply multiple parameters 
df_mod.groupby(['Type 1', 'Type 2']).count()['count']

Type 1  Type 2  
Bug     Electric     2
        Fighting     2
        Fire         2
        Flying      14
        Ghost        1
                    ..
Water   Ice          3
        Poison       3
        Psychic      5
        Rock         4
        Steel        1
Name: count, Length: 136, dtype: int64

In [23]:
##Read chunk size data when data is too big
##For example read 5 rows at a time

# for df_mod in pd.read_csv('modified_new.csv', chunksize=5):
#     print("Chunk Df")
#     print(df)

### PCA - Principal Component Analysis 
#### It is a linear dimensionality reduction technique. 

### Why PCA ?
####  Because there are a large number of variables or dimensions along which the data is distributed, visualization can be a challenge. Hence, PCA can do that for you since it projects the data into a lower dimension, thereby allowing you to visualize the data in a 2D or 3D space with a naked eye.

In [24]:
##Principal Component Analysis(PCA) used for dimension reduction of dataset

import matplotlib.pyplot as plt
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

%matplotlib inline

#### Selecting numeric columns in which PSA technique has to apply to reduce dimension 

In [25]:
pca_col =['HP','Attack', 'Defense', 'Sp. Atk', 'Sp. Def', 'Speed']

#### Standardizing the data since PCA's output is influenced based on the scale of the features of the data.

In [26]:
scaler = StandardScaler()
scaler.fit(df_mod[pca_col])

StandardScaler()

In [27]:
scaled_data = scaler.transform(df_mod[pca_col])

#### While applying StandardScaler, each feature of your data should be normally distributed such that it will scale the distribution to a mean of zero and a standard deviation of one.

In [28]:
np.mean(scaled_data), np.std(scaled_data)

(-5.033011044967376e-17, 1.0)

#### It has been converted into array and this is based on Standard Deviation is 1 and Mean = 0.


In [29]:
scaled_data

array([[-0.95062622, -0.92490618, -0.79715364, -0.2391303 , -0.24818854,
        -0.80150318],
       [-0.3628221 , -0.5241302 , -0.34791742,  0.21955954,  0.29115635,
        -0.28501525],
       [ 0.42091674,  0.09244823,  0.29384863,  0.83114599,  1.01028289,
         0.40363531],
       ...,
       [ 0.42091674,  0.95565803, -0.44418232,  2.36011211,  2.08897269,
         0.05931003],
       [ 0.42091674,  2.49710411, -0.44418232,  2.97169856,  2.08897269,
         0.40363531],
       [ 0.42091674,  0.95565803,  1.4811158 ,  1.74852566,  0.65071962,
         0.05931003]])

#### Let's convert the normalized features into a tabular format with the help of DataFrame

In [30]:
feature_cols = ['feature'+str(i) for i in range(scaled_data.shape[1])]
normalised_pokemon = pd.DataFrame(scaled_data,columns=feature_cols)
normalised_pokemon.tail()

feature0  feature1  feature2  feature3  feature4  feature5
795 -0.754692  0.647369  2.443765  0.831146  2.808099 -0.629341
796 -0.754692  2.497104  1.160233  2.665905  1.369846  1.436611
797  0.420917  0.955658 -0.444182  2.360112  2.088973  0.059310
798  0.420917  2.497104 -0.444182  2.971699  2.088973  0.403635
799  0.420917  0.955658  1.481116  1.748526  0.650720  0.059310

### Use the sklearn library to import the PCA module

In [31]:
from sklearn.decomposition import PCA

#### Projecting the six-dimensional data to two-dimensional principal components. You will pass the number of components (n_components=0.6) and finally call fit_transform on the aggregate data.

In [32]:
pca = PCA(n_components = 0.6)
principal_components =pca.fit(scaled_data)

In [33]:
x_pca = pca.transform(scaled_data)
scaled_data.shape

(800, 6)

#### Shape of dataframe after PCA reduction technique.

In [34]:
x_pca.shape

(800, 2)

### Let's create a DataFrame that will have the principal component values for all 800 samples.

In [35]:
principal_comp_df = pd.DataFrame(data = x_pca, columns = ['Principal_comp_1', 'Principal_comp_2'])
principal_comp_df.head(4)

Principal_comp_1  Principal_comp_2
0         -1.556375         -0.021482
1         -0.362867         -0.050269
2          1.280152         -0.062720
3          2.620916          0.704263

### Now, find the explained_variance_ratio. It will provide you with the amount of information or variance each principal component holds after projecting the data to a lower dimensional subspace.

In [36]:
print('Explained variation per principal component: {}'.format(pca.explained_variance_ratio_))

Explained variation per principal component: [0.45190665 0.18225358]


#### From above result you can observe that the principal component 1 holds 45.2% of the information while the principal component 2 holds only 18.2% of the information and in the process of reduction 36.6% information was lost.